File: java/BestTimetoBuyandSellStock.java
/*
 * Filename: BestTimetoBuyandSellStock.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 29 Jun 2015 09:37:39 AM EDT
 */

public class Solution {
	public int maxProfit(int[] prices) {
		if (prices == null || prices.length <= 1) {
			return 0;
		}

		if (prices.length == 2) {
			return Math.max(0, prices[1] - prices[0]);
		}

		int min = Integer.MAX_VALUE;
		int max = 0;
		int diff;

		for (int i = 0; i < prices.length; i++) {
			min = Math.min(min, prices[i]);
			max = Math.max(max, prices[i] - min);
		}

		return max;
	}
}

File: cpp/best_time_to_buy_and_sell_stock.cc
/*
 * Filename: best_time_to_buy_and_sell_stock.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 05 Aug 2015 06:40:12 PM EDT
 */

#include <iostream>
#include <climits>

using namespace std;

class Solution {
	public:
		int maxProfit(vector<int>& prices)
		{
			if (prices.empty())
			{
				return 0;
			}

			int mini = INT_MAX;
			int res = 0;

			for (vector<int>::size_type i = 0; i < prices.size(); i++)
			{
				mini = min(mini, prices[i]);
				res = max(res, prices[i] - mini);
			}

			return res;
		}
};
